io.openems.edge.controller.api.mqtt/src/io/openems/edge/controller/api/mqtt/ControllerApiMqttImpl.java

```diff
@@ -67,35 +73,72 @@ public ControllerApiMqttImpl() {
 				Controller.ChannelId.values(), //
 				ControllerApiMqtt.ChannelId.values() //
 		);
+		this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
 	}
 
 	@Activate
-	private void activate(ComponentContext context, Config config) throws Exception {
+	protected void activate(ComponentContext context, Config config) {
 		this.config = config;
-
-		// Publish MQTT messages under the topic "edge/edge0/..."
 		this.topicPrefix = String.format(ControllerApiMqtt.TOPIC_PREFIX, config.clientId());
 
 		super.activate(context, config.id(), config.alias(), config.enabled());
-		this.mqttConnector.connect(config.uri(), config.clientId(), config.username(), config.password(),
-				config.certPem(), config.privateKeyPem(), config.trustStorePem()).thenAccept(client -> {
-					this.mqttClient = client;
-					this.logInfo(this.log, "Connected to MQTT Broker [" + config.uri() + "]");
-				});
+		scheduleReconnect();
+	}
+
+	@Modified
+	protected void modified(ComponentContext context, Config config) {
+		this.config = config;
+		this.topicPrefix = String.format(ControllerApiMqtt.TOPIC_PREFIX, config.clientId());
+
+		this.logInfo(this.log, "Configuration modified. Adjusting MQTT connection settings.");
+
+		scheduleReconnect();
+	}
+
+	private void scheduleReconnect() {
+		if (this.scheduledExecutorService != null && !this.scheduledExecutorService.isShutdown()) {
```

Comment:
It's usually preferable to not implement `modified()` because it requires complicated logic like this. If you choose the adaptive approach described above, maybe `activate()` is sufficient?

Restarting the thread is only required if the `reconnectionAttemptInterval` changes. You could validate this here, instead of always restarting the executor.

Also you should be able to use `ThreadPoolUtils.shutdownAndAwaitTermination` here as well, right?